network/src/peers/peer_book.rs

```diff
@@ -206,40 +207,40 @@ impl PeerBook {
         };
 
         // Remove the peer's address from the list of connecting peers.
-        self.connecting_peers.write().remove(&address);
+        self.connecting_peers.remove(address).await;
 
         metrics::decrement_gauge!(stats::CONNECTIONS_CONNECTING, 1.0);
 
         // Update the peer info to connected.
         peer_info.set_connected();
 
         // Add the address into the connected peers.
-        self.connected_peers.write().insert(listener, peer_info);
-
+        self.connected_peers.insert(listener, peer_info).await;
+        // On success, increment the connected peer count.
         metrics::increment_gauge!(stats::CONNECTIONS_CONNECTED, 1.0);
     }
 
     ///
     /// Removes the given address from the connecting and connected peers in this `PeerBook`,
     /// and adds the given address to the disconnected peers in this `PeerBook`.
     ///
-    pub fn set_disconnected(&self, address: SocketAddr) -> bool {
+    pub async fn set_disconnected(&self, address: SocketAddr) -> bool {
         // Case 1 - The given address is a connecting peer, attempt to disconnect.
-        if self.connecting_peers.write().remove(&address) {
+        if self.connecting_peers.remove(address).await.is_some() {
             metrics::decrement_gauge!(stats::CONNECTIONS_CONNECTING, 1.0);
-            return false;
+            return true;
```

Comment:
```suggestion
            return false;
```
this should still be `false`, as it's only used to indicate whether the disconnected peer had been `connected`, and not `connecting`